internal/mock/alloydb.go

```diff
@@ -241,3 +252,70 @@ func StartServerProxy(t *testing.T, inst FakeAlloyDBInstance) func() {
 		ln.Close()
 	}
 }
+
+// metadataExchange mimics server side behavior in four steps:
+//
+//  1. Read a big endian uint32 (4 bytes) from the client. This is the number of
+//     bytes the message consumes. The length does not include the initial four
+//     bytes.
+//
+//  2. Read the message from the client using the message length and unmarshal
+//     it into a MetadataExchangeResponse message.
+//
+// The real server implementation will then validate the client has connection
+// permissions using the provided OAuth2 token based on the auth type. Here in
+// the test implementation, the server does nothing.
+//
+//  3. Prepare a response and write the size of the response as a uint32 (4
+//     bytes)
+//
+// 4. Marshal the response to bytes and write those to the client as well.
+//
+// Subsequent interactions with the test server use the database protocol.
+func metadataExchange(conn net.Conn) error {
+	msgSize := make([]byte, 4)
+	n, err := conn.Read(msgSize)
+	if err != nil {
+		return err
+	}
+	if n != 4 {
+		return fmt.Errorf("read %d bytes, want = 4", n)
+	}
+
+	size := binary.BigEndian.Uint32(msgSize)
+	buf := make([]byte, size)
+	n, err = conn.Read(buf)
+	if err != nil {
+		return err
+	}
+	if n != int(size) {
+		return fmt.Errorf("read %d bytes, want = %d", n, size)
+	}
+
+	m := &connectorspb.MetadataExchangeRequest{}
+	err = proto.Unmarshal(buf, m)
+	if err != nil {
+		return err
+	}
+
+	resp := &connectorspb.MetadataExchangeResponse{
```

Comment:
Yes, good idea. This will require some restructing of the test code. Let's do that in a separate PR. Meanwhile, thinking about error paths made me realize we weren't closing the TLS connection when the metadata exchange protocol fails. The test cleanup will help put an assertion around this.

Reply:
Filed https://github.com/GoogleCloudPlatform/alloydb-go-connector/issues/427 for this work.